docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,312 @@
+  ``dependency_tracker`` object is used while scheduling and cleaning up bears
+  and also while finishing bear tasks.
+- **Task Scheduling:** The bears that have no dependencies or whose
+  dependencies have been resolved, only their tasks will be scheduled for
+  execution. Before executing any task coala looks it up in the cache.
+  In case of a hit, the existing results that are stored in the cache for the
+  corresponding task arguments are called using ``execute_task_with_cache``
+  method. In case of a miss or if coala is run without a cache the task is
+  executed. The bears without any running tasks are cleaned up from the state of
+  an ongoing run by resolving its dependencies, scheduling dependent bears
+  and removing the bear from the ``running_tasks`` dict. When a task belonging
+  to a particular bear completes, then ``_finish_task`` method is responsible
+  for checking whether the bear execution is complete and the result generated
+  from the task have been returned. If any new tasks are to be scheduled then
+  this method is responsible for doing so after the dependencies have been
+  resolved.
+- **result_callback :** Each of the bears scheduled to run for a session has
+  this callback function which is called when the results are available. It
+  should have the following signature ``def result_callback(result)``.
+- **executors:** These are instances of the ``ProcessPoolExecutor`` class
+  provided by the ``concurrent.futures`` module that is used to execute function
+  calls asynchronously using a pool of processes. Custom ``executor`` instances
+  can be passed to functions as parameters to run the bears. If not provided
+  then a ``ProcessPoolExecutor`` is used using the maximum number of processors
+  available on the machine.
+
+Custom parallelization for bears
+--------------------------------
+A bear inheriting from the class FileBear can parallelize tasks for each file
+given. A bear inheriting from the class DependencyBear can parallelize tasks for
+each dependency result. A bear inheriting from the class ProjectBear does not
+parallelize tasks for each file as it runs on the whole codebase given.
+
+Superior Caching
+----------------
+
+The NextGen-Core's caching mechanism is based on task objects. Bears can offload
+tasks via `generate_tasks()` which get executed by a Python pool. Structure wise
+the cache is a dictionary-like-object with bear types and cache-tables as key
+value pairs. The cache-tables themselves are dictionary-like-objects that map
+the hash values of the task objects (generated by
+``PersistentHash.persistent_hash``) to the bear results.
+
+At the time of scheduling the bears, the core performs a cache lookup. If the
+parameters to ``execute_tasks()`` are the same (in other words it looks for
+identical task objects in the cache, and fetches their corresponding results
+if found) as that of the previous run then instead of executing that bear again
+we get the cached results of that bear.
```

Comment:
One thing we definitely should write (not sure exactly where, but I guess this is definitely caching related):

The coala core expects analysis functions (i.e. `analyze`) to provide results that only depend on the input parameters. So `analyze` shall be a mapping from `section, file_dict -> results`, or better: a mapping from `analyze-parameters -> results`. Using volatile values like randomness or time-dependent data without putting it into the task objects is prohibited (we can't check that though, it's Python, you are free to do anything you like. But we are all adults here, right? :P). Not respecting this causes unknown behaviour in coala.